src/experimental-layers/src/solid-polygon-layer/polygon-tesselator.js

```diff
@@ -103,109 +139,112 @@ function getPolygonOffsets(polygons) {
   return offsets;
 }
 
-// Returns the offset of each hole polygon in the flattened array for that polygon
-function getHoleIndices(complexPolygon) {
-  let holeIndices = null;
-  if (complexPolygon.length > 1) {
-    let polygonStartIndex = 0;
-    holeIndices = [];
-    complexPolygon.forEach(polygon => {
-      polygonStartIndex += polygon.length;
-      holeIndices.push(polygonStartIndex);
-    });
-    // Last element points to end of the flat array, remove it
-    holeIndices.pop();
-  }
-  return holeIndices;
-}
-
 function calculateIndices({polygons, IndexType = Uint32Array}) {
   // Calculate length of index array (3 * number of triangles)
   const indexCount = 3 * getTriangleCount(polygons);
   const offsets = getPolygonOffsets(polygons);
 
   // Allocate the attribute
-  // TODO it's not the index count but the vertex count that must be checked
-  if (IndexType === Uint16Array && indexCount > 65535) {
-    throw new Error('Vertex count exceeds browser\'s limit');
-  }
   const attribute = new IndexType(indexCount);
 
   // 1. get triangulated indices for the internal areas
   // 2. offset them by the number of indices in previous polygons
   let i = 0;
   polygons.forEach((polygon, polygonIndex) => {
-    for (const index of calculateSurfaceIndices(polygon)) {
+    for (const index of Polygon.getSurfaceIndices(polygon)) {
       attribute[i++] = index + offsets[polygonIndex];
     }
   });
 
   return attribute;
 }
 
-/*
- * Get vertex indices for drawing complexPolygon mesh
- * @private
- * @param {[Number,Number,Number][][]} complexPolygon
- * @returns {[Number]} indices
- */
-function calculateSurfaceIndices(complexPolygon) {
-  // Prepare an array of hole indices as expected by earcut
-  const holeIndices = getHoleIndices(complexPolygon);
-  // Flatten the polygon as expected by earcut
-  const verts = flattenVertices(complexPolygon);
-  // Let earcut triangulate the polygon
-  return earcut(verts, holeIndices, 3);
-}
-
-function calculatePositions({polygons, pointCount, fp64}) {
+function updatePositions(
+  {positions, positions64xyLow, nextPositions, nextPositions64xyLow},
+  {polygons, extruded, fp64}
+) {
   // Flatten out all the vertices of all the sub subPolygons
-  const attribute = new Float32Array(pointCount * 3);
-  let attributeLow;
-  if (fp64) {
-    // We only need x, y component
-    attributeLow = new Float32Array(pointCount * 2);
-  }
   let i = 0;
   let j = 0;
-  for (const polygon of polygons) {
-    Polygon.forEachVertex(polygon, vertex => { // eslint-disable-line
+  let nextI = 0;
+  let nextJ = 0;
+  let startVertex = null;
+
+  const popStartVertex = () => {
+    if (startVertex && extruded) {
+      nextPositions[nextI++] = startVertex.x;
+      nextPositions[nextI++] = startVertex.y;
+      nextPositions[nextI++] = startVertex.z;
+      if (fp64) {
```

Comment:
Idle thought... If we always calculated the fp64 attributes, we could get rid of some annoying logic when switching modes. Wonder what the perf impact is, probably small.